src/activity_actor.cpp

```diff
@@ -2597,6 +2603,487 @@ std::unique_ptr<activity_actor> milk_activity_actor::deserialize( JsonIn &jsin )
     return actor.clone();
 }
 
+void read_activity_actor::start( player_activity &act, Character &who )
+{
+    if( !book->is_book() ) {
+        debugmsg( "ACT_READ on a non-book item" );
+        act.set_to_null();
+        return;
+    }
+
+    // make reader wield book to prevent
+    // losing item_location errors while loading
+    if( !who.is_wielding( *book ) ) {
+        if( !who.is_armed() ||
+            ( who.is_npc() || query_yn( _( "Wield the %s and start reading?" ), book->tname() ) ) )   {
+
+            if( !who.wield( *book ) ) {
+                add_msg( m_bad, "%s", who.can_wield( *book ).c_str() );
+                act.set_to_null();
+                return;
+            }
+            // location changed
+            book = item_location( who, &who.weapon );
+        } else {
+            act.set_to_null();
+            return;
+        }
+    }
+
+    bktype = book->type->use_methods.count( "MA_MANUAL" ) ?
+             book_type::martial_art : book_type::normal;
+
+    // push copy of book for focus calculation
+    // avatar::update_mental_focus and avatar::calc_focus_equilibrium
+    act.targets.push_back( book );
+
+    act.moves_total = moves_total;
+    act.moves_left = moves_total;
+}
+
+void read_activity_actor::do_turn( player_activity &act, Character &who )
+{
+    if( who.fine_detail_vision_mod() > 4 ) {
+        //It got too dark during the process of reading, bail out.
+        act.set_to_null();
+        who.add_msg_if_player( m_bad, _( "It's too dark to read!" ) );
+        return;
+    }
+
+    if( !bktype.has_value() ) {
+        bktype = book->type->use_methods.count( "MA_MANUAL" ) ?
+                 book_type::martial_art : book_type::normal;
+    }
+
+    if( who.is_avatar() ) {
+        if( bktype.value() == book_type::martial_art && one_in( 3 ) ) {
+            who.mod_stamina( -1 );
+        }
+    } else {
+        who.moves = 0;
+    }
+}
+
+void read_activity_actor::read_book( Character &learner,
+                                     const cata::value_ptr<islot_book> &islotbook,
+                                     SkillLevel &skill_level, const double penalty )
+{
+    const int originalSkillLevel = skill_level.level();
+
+    // Calculate experience gained
+    /** @EFFECT_INT increases reading comprehension */
+    // Enhanced Memory Banks modestly boosts experience
+    int min_ex = std::max( 1, islotbook->time / 10 + learner.get_int() / 4 );
+    int max_ex = islotbook->time / 5 + learner.get_int() / 2 - originalSkillLevel;
+
+    if( learner.has_active_bionic( bio_memory ) ) {
+        min_ex += 2;
+    }
+
+    min_ex = learner.adjust_for_focus( min_ex );
+    max_ex = learner.adjust_for_focus( max_ex );
+
+    if( max_ex < 2 ) {
+        max_ex = 2;
+    }
+
+    if( max_ex > 10 ) {
+        max_ex = 10;
+    }
+
+    if( max_ex < min_ex ) {
+        max_ex = min_ex;
+    }
```

Comment:
Can be simplified with
```suggestion
    max_ex = clamp( max_ex, 2, 10 );
    max_ex = std::max( min_ex, max_ex );
```